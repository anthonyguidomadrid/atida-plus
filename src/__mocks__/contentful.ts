export const getEntries = jest.fn().mockResolvedValue({})

export const createClient = jest.fn().mockImplementation(() => ({
  getEntries
}))
