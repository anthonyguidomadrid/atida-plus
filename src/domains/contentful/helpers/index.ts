export * from './check-if-graphql-response'
