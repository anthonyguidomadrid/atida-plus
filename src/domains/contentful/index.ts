export * from './normalizers'
export * from './component-mappers'
export * from './helpers'
