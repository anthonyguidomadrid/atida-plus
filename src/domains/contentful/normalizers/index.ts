export * from './asset'
export * from './brand'
export * from './category'
export * from './color'
export * from './content-block-with-image'
export * from './expert-signature'
export * from './recommendation'
export * from './slider'
export * from './hero-banner'
export * from './icon'
export * from './link'
export * from './menu'
export * from './rich-text'
export * from './hero-banner'
export * from './category-grid'
export * from './usps-card'
export * from './seo'
export * from './static-content-block'
export * from './static-header-block'
export * from './static-recommendation-block'
export * from './container-of-content-blocks'
export * from './group-of-static-content-blocks'
export * from './marketing-teaser'
export * from './category-overview-page'
export * from './promotion'
export * from './top-brands'
export * from './campaign-hero-banner'
export * from './link-block'
export * from './organization'
export * from './voucher-codes'
