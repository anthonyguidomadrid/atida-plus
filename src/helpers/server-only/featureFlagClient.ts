import LaunchDarkly from 'launchdarkly-node-server-sdk'
import getConfig from 'next/config'
import { FeatureFlagValue } from '~components/helpers/FeatureFlags/hooks'
import {
  FeatureFlag,
  allPagesFeatureFlags
} from '~config/constants/feature-flags'
import { getCountryFromLocale, getIso2CodeFromLocale } from '~helpers'

let launchDarklyClient: LaunchDarkly.LDClient

async function initialize() {
  const logger = LaunchDarkly.basicLogger({
    level: 'error'
  })
  const options: LaunchDarkly.LDOptions = { logger: logger }

  const client = LaunchDarkly.init(
    process.env.LAUNCH_DARKLY_SDK_KEY as string,
    options
  )
  await client.waitForInitialization()
  return client
}

async function getLaunchDarklyClient(): Promise<LaunchDarkly.LDClient> {
  if (launchDarklyClient) return launchDarklyClient
  return (launchDarklyClient = await initialize())
}

export const loadFeatureFlags = async (
  locale: string | undefined,
  pageFeatureFlags?: FeatureFlag[],
  user?: {
    key: string
    country?: string
    custom?: {
      store?: string
    }
  }
): Promise<Record<string, FeatureFlagValue>> => {
  const client = await getLaunchDarklyClient()
  const { publicRuntimeConfig } = getConfig()
  const defaultUser = {
    key: '',
    country: getCountryFromLocale(locale),
    custom: {
      store: `${publicRuntimeConfig.brand ?? 'ATIDA'} ${getIso2CodeFromLocale(
        locale
      )}`
    }
  }
  pageFeatureFlags = pageFeatureFlags ?? []

  return Array.from(
    new Set([...allPagesFeatureFlags, ...pageFeatureFlags])
  ).reduce(async (featureFlags, flagName) => {
    return {
      ...(await featureFlags),
      [flagName]: await client.variation(
        flagName,
        {
          ...defaultUser,
          ...user
        },
        false
      )
    }
  }, {})
}

export const loadFeatureFlag = async (
  locale: string | undefined,
  featureFlag: FeatureFlag,
  user?: {
    key: string
    country?: string
    custom?: {
      store?: string
    }
  }
): Promise<FeatureFlagValue> => {
  const client = await getLaunchDarklyClient()
  const { publicRuntimeConfig } = getConfig()

  const defaultUser = {
    key: '',
    country: getCountryFromLocale(locale),
    custom: {
      store: `${publicRuntimeConfig.brand ?? 'ATIDA'} ${getIso2CodeFromLocale(
        locale
      )}`
    }
  }

  return client.variation(
    featureFlag,
    {
      ...defaultUser,
      ...user
    },
    false
  )
}
