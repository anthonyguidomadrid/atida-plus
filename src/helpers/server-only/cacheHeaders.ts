import getConfig from 'next/config'
import { FeatureFlag } from '~config/constants/feature-flags'
import { loadFeatureFlag } from './featureFlagClient'

const DEFAULT_S_MAX_AGE = 1
const DEFAULT_MAX_AGE = 3

export const getPageCacheHeader = async (
  locale?: string,
  isPublicPage?: boolean
): Promise<string> => {
  const { serverRuntimeConfig } = getConfig()

  if (
    !serverRuntimeConfig.cacheControlProfile ||
    serverRuntimeConfig.cacheControlProfile === 'private'
  )
    return "private, stale-if-error=0, no-cache='Set-Cookie'"

  if (isPublicPage) {
    const newCacheValues = await loadFeatureFlag(
      locale,
      FeatureFlag.CACHING_NEW_VALUES
    )

    return `public, no-cache='Set-Cookie', s-maxage=${DEFAULT_S_MAX_AGE}${
      newCacheValues ? `, max-age=${DEFAULT_MAX_AGE}, must-revalidate` : ''
    }`
  }

  return 'no-cache, no-store, must-revalidate'
}
