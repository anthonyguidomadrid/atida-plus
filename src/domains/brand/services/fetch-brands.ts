import {
  createClient,
  transformLocaleToUppercase,
  getCountryTagFromLocale
} from '~helpers'
import { BrandPagesQuery, BrandPages } from '~generated-graphql'
import getConfig from 'next/config'
import {
  normalizeBrandPage,
  normalizeBrandData,
  Brand,
  Brands,
  ContentfulBrand,
  normalizeBrandRest
} from '~domains/contentful/normalizers/brand'
import { cache } from '~helpers/server-only/redisClient'
import { createContentfulClient } from '~domains/contentful/helpers/client'
import { createLocalClient } from '~domains/contentful/helpers/contentful-api'
import { ContentfulPage } from '~domains/page'
import { ContentfulClientApi } from 'contentful'
import { FeatureFlag } from '~config/constants/feature-flags'
import { loadFeatureFlag } from '~helpers/server-only/featureFlagClient'
import {
  getDefaultLocale,
  getAvailableLocales
} from '~domains/translated-routes'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

export const fetchBrands = async (locale?: string): Promise<Brands> => {
  const useCDA = await loadFeatureFlag(
    locale,
    FeatureFlag.CONTENTFUL_FETCH_BRANDS_WITH_CDA
  )

  if (useCDA) return fetchBrandsWithCDA(locale)

  const { serverRuntimeConfig } = getConfig()
  const client = createClient({
    options: {
      baseURL: `${serverRuntimeConfig.cmsGraphQLHost}/spaces/${serverRuntimeConfig.cmsSpaceId}/environments/${serverRuntimeConfig.cmsEnvironmentId}`,
      headers: {
        Authorization: `Bearer ${serverRuntimeConfig.cmsToken}`
      },
      adapter: cache?.adapter
    },
    locale,
    interceptorOptions: {
      mode: 'contentful-graphql'
    }
  })

  const brandsResponse = await client.post<{ data: BrandPagesQuery }>('', {
    query: BrandPages.loc?.source.body,
    variables: {
      allLocales: getAvailableLocales().map(transformLocaleToUppercase),
      locale: transformLocaleToUppercase(locale),
      tags: getCountryTagFromLocale(locale)
    }
  })

  return normalizeBrandData({
    data: brandsResponse?.data?.data?.pageCollection ?? undefined,
    includes:
      brandsResponse?.data?.data?.pageCollection?.items &&
      brandsResponse?.data?.data?.pageCollection?.items
        .map(page => {
          if (page?.referencedContent?.__typename === 'Brand') {
            //@ts-ignore
            return normalizeBrandPage(page?.referencedContent, page?.slug)
          }
          return
        })
        .filter((page: unknown): page is Brand => !!page)
  })
}

const PAGE_SIZE = 500

const brandFetcher = async (
  locale: string,
  client: ContentfulClientApi,
  skip = 0
) =>
  createLocalClient(client).getContentfulEntries<ContentfulPage>(
    ['page'],
    locale,
    1,
    skip,
    PAGE_SIZE,
    'pageType[in]',
    'Brand'
  )

const fetchBrandsWithCDA = async (
  locale = getDefaultLocale()
): Promise<Brands> => {
  const client = createContentfulClient()
  const response = await brandFetcher(locale, client)
  let responses = [response]

  if (response.total > PAGE_SIZE) {
    const leftToFetch = response.total - PAGE_SIZE
    const remainingPages = Math.ceil(leftToFetch / PAGE_SIZE)
    const pageRequests = new Array(remainingPages)
      .fill(0)
      .map((_value, idx) => brandFetcher(locale, client, PAGE_SIZE * (idx + 1)))
    const pageResponses = await Promise.all(pageRequests)
    responses = [response, ...pageResponses]
  }

  const items = responses.reduce((acc, curr) => {
    const currentItems = curr?.items
      ? (curr?.items
          .map(i =>
            normalizeBrandRest(
              i.fields?.referencedContent as ContentfulBrand,
              i.fields?.slug
            )
          )
          .filter(item => item) as Partial<Brand>[])
      : []
    return [...acc, ...currentItems]
  }, [] as Brand[])

  return {
    total: response.total,
    items
  }
}
