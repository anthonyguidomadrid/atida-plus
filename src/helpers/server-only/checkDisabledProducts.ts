import { removeFromBasket } from '~domains/basket/services/remove-from-basket'
import { BasketWithProducts } from '~domains/basket/types'
import { fetchBasket } from '~domains/basket/services/fetch-basket'

export const checkDisabledProducts = async (
  basket: BasketWithProducts,
  locale: string | string[],
  anonymousCustomerUniqueId: string,
  token: string
): Promise<BasketWithProducts> => {
  const disabledItem = basket.items?.find(item => !item.product?.enabled)
  if (disabledItem) {
    await removeFromBasket(
      locale.toString(),
      { ...disabledItem, basketId: basket.id as string },
      {
        anonymousCustomerUniqueId,
        token
      }
    )
    basket = await fetchBasket(locale.toString(), {
      anonymousCustomerUniqueId,
      token
    })
    return checkDisabledProducts(
      basket,
      locale,
      anonymousCustomerUniqueId,
      token
    )
  }

  return basket
}
