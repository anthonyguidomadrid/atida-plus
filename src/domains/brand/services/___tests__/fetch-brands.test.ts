/**
 * @jest-environment node
 */
// @ts-ignore
import { getEntries } from 'contentful'
import { loadFeatureFlag } from '~helpers/server-only/featureFlagClient'
import axios from 'axios'
import { fetchBrands } from '../fetch-brands'

const contentfulBrands = {
  data: {
    pageCollection: {
      total: 1,
      items: [
        {
          __typename: 'Page',
          referencedContent: {
            __typename: 'Brand',
            title: 'Babaria',
            id: 'babaria-test-brand',
            image: {
              __typename: 'Asset',
              url:
                'https://images.ctfassets.net/7g2w796onies/38LA8ERIEBglSvG3XR2vCi/01fbdbc080c12381cf118fd8c90d1bf4/babaria-body-milk.jpeg'
            },
            url: {
              pageCollection: {
                items: [
                  {
                    slug: 'marca/babaria',
                    referenceContent: {
                      id: 'babaria-test-brand'
                    }
                  }
                ]
              }
            }
          }
        }
      ]
    }
  }
}

describe(fetchBrands, () => {
  beforeEach(() => {
    getEntries.mockResolvedValue({
      total: 3,
      limit: 500,
      skip: 0,
      items: [
        {
          fields: {
            title: 'Mercks',
            slug: 'marca/mercks',
            pageType: 'Brand',
            referencedContent: { fields: { title: 'Merck', id: 'mercks' } }
          }
        },
        {
          fields: {
            title: 'Patata',
            slug: 'marca/patata',
            pageType: 'Brand',
            referencedContent: { fields: { title: 'Patata', id: 'patata' } }
          }
        },
        {
          fields: {
            title: 'Atida',
            slug: 'marca/atida',
            pageType: 'Brand',
            referencedContent: { fields: { title: 'Atida', id: 'atida' } }
          }
        }
      ]
    })
  })
  describe('fetch brands with Content Delivery API', () => {
    beforeEach(() => {
      ;(loadFeatureFlag as jest.Mock).mockResolvedValue(true)
    })

    afterEach(() => {
      ;(loadFeatureFlag as jest.Mock).mockReset()
    })

    it('uses the CDA when the FF is on', async () => {
      await fetchBrands('en-gb')
      expect(getEntries).toHaveBeenCalledTimes(1)
      expect(getEntries).toHaveBeenCalledWith({
        include: 1,
        locale: 'en-GB',
        limit: 500,
        skip: 0,
        content_type: 'page',
        'metadata.tags.sys.id[in]': 'country-gb',
        'fields.pageType[in]': 'Brand'
      })
    })

    it('generates a brand per item in the response', async () => {
      const result = await fetchBrands('en-gb')
      expect(result.total).toBe(3)
      expect(result.items?.length).toBe(3)
    })

    it('generates a correct URL for a brand', async () => {
      const result = await fetchBrands('en-gb')
      expect(result.items?.[0].url).toBe('/marca/mercks')
    })
  })

  describe('calls Content Delivery API with pagination', () => {
    beforeEach(() => {
      ;(loadFeatureFlag as jest.Mock).mockResolvedValue(true)
    })

    afterEach(() => {
      ;(loadFeatureFlag as jest.Mock).mockReset()
    })

    it('can fetch more than one page', async () => {
      getEntries.mockResolvedValueOnce({
        total: 1800,
        limit: 500,
        skip: 0,
        items: [
          {
            fields: {
              title: 'Mercks',
              slug: 'marca/mercks',
              pageType: 'Brand',
              referencedContent: { fields: { title: 'Merck', id: 'mercks' } }
            }
          }
        ]
      })
      getEntries.mockResolvedValueOnce({
        total: 1800,
        limit: 500,
        skip: 500,
        items: [
          {
            fields: {
              title: 'Patata',
              slug: 'marca/patata',
              pageType: 'Brand',
              referencedContent: { fields: { title: 'Patata', id: 'patata' } }
            }
          }
        ]
      })
      getEntries.mockResolvedValueOnce({
        total: 1800,
        limit: 500,
        skip: 1000,
        items: [
          {
            fields: {
              title: 'Atida',
              slug: 'marca/atida',
              pageType: 'Brand',
              referencedContent: { fields: { title: 'Atida', id: 'atida' } }
            }
          }
        ]
      })
      getEntries.mockResolvedValueOnce({
        total: 1800,
        limit: 500,
        skip: 1500,
        items: [
          {
            fields: {
              title: 'Test',
              slug: 'marca/test',
              pageType: 'Brand',
              referencedContent: { fields: { title: 'Test', id: 'test' } }
            }
          }
        ]
      })

      const response = await fetchBrands('en-gb')
      expect(getEntries).toHaveBeenCalledTimes(4)
      expect(getEntries).toHaveBeenNthCalledWith(1, {
        include: 1,
        locale: 'en-GB',
        limit: 500,
        skip: 0,
        content_type: 'page',
        'metadata.tags.sys.id[in]': 'country-gb',
        'fields.pageType[in]': 'Brand'
      })
      expect(getEntries).toHaveBeenNthCalledWith(2, {
        include: 1,
        locale: 'en-GB',
        limit: 500,
        skip: 500,
        content_type: 'page',
        'metadata.tags.sys.id[in]': 'country-gb',
        'fields.pageType[in]': 'Brand'
      })
      expect(getEntries).toHaveBeenNthCalledWith(3, {
        include: 1,
        locale: 'en-GB',
        limit: 500,
        skip: 1000,
        content_type: 'page',
        'metadata.tags.sys.id[in]': 'country-gb',
        'fields.pageType[in]': 'Brand'
      })
      expect(getEntries).toHaveBeenNthCalledWith(4, {
        include: 1,
        locale: 'en-GB',
        limit: 500,
        skip: 1500,
        content_type: 'page',
        'metadata.tags.sys.id[in]': 'country-gb',
        'fields.pageType[in]': 'Brand'
      })
      expect(response).toEqual({
        items: [
          {
            id: 'mercks',
            title: 'Merck',
            url: '/marca/mercks'
          },
          {
            id: 'patata',
            title: 'Patata',
            url: '/marca/patata'
          },
          {
            id: 'atida',
            title: 'Atida',
            url: '/marca/atida'
          },
          {
            id: 'test',
            title: 'Test',
            url: '/marca/test'
          }
        ],
        total: 1800
      })
    })

    it('can fetch a single page', async () => {
      getEntries.mockResolvedValueOnce({
        total: 499,
        limit: 500,
        skip: 0,
        items: [
          {
            fields: {
              title: 'Atida',
              slug: 'marca/atida',
              pageType: 'Brand',
              referencedContent: { fields: { title: 'Atida', id: 'atida' } }
            }
          }
        ]
      })
      const response = await fetchBrands('en-gb')
      expect(getEntries).toHaveBeenCalledTimes(1)
      expect(getEntries).toHaveBeenNthCalledWith(1, {
        include: 1,
        locale: 'en-GB',
        limit: 500,
        skip: 0,
        content_type: 'page',
        'metadata.tags.sys.id[in]': 'country-gb',
        'fields.pageType[in]': 'Brand'
      })
      expect(response).toEqual({
        items: [
          {
            id: 'atida',
            title: 'Atida',
            url: '/marca/atida'
          }
        ],
        total: 499
      })
    })
  })
  describe('fetch brands with GrahpQL', () => {
    beforeEach(() => {
      ;(loadFeatureFlag as jest.Mock).mockResolvedValue(false)
    })

    afterEach(() => {
      ;(loadFeatureFlag as jest.Mock).mockReset()
    })
    it('does not use the CDA when the FF is on', async () => {
      await fetchBrands('en-gb')
      expect(getEntries).toHaveBeenCalledTimes(0)
    })
    it('creates the client passes the locale to request', async () => {
      ;(axios.post as jest.Mock).mockResolvedValue({
        data: contentfulBrands
      })

      await fetchBrands('en-gb')
      expect(axios.post).toHaveBeenLastCalledWith('', {
        query: expect.any(String),
        variables: {
          locale: 'en-GB',
          tags: 'country-gb',
          allLocales: ['en-GB', 'pt-PT', 'es-ES', 'de-DE']
        }
      })
    })

    it('returns the normalized response', async () => {
      ;(axios.post as jest.Mock).mockResolvedValueOnce({
        data: contentfulBrands
      })
      const response = await fetchBrands('en-gb')
      expect(response).toMatchSnapshot()
    })
  })
})
