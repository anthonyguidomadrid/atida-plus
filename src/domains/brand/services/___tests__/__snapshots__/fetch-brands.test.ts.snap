// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchBrands fetch brands with GrahpQL returns the normalized response 1`] = `
Object {
  "items": Array [
    Object {
      "id": "babaria-test-brand",
      "image": Object {
        "url": "https://images.ctfassets.net/7g2w796onies/38LA8ERIEBglSvG3XR2vCi/01fbdbc080c12381cf118fd8c90d1bf4/babaria-body-milk.jpeg",
      },
      "title": "Babaria",
    },
  ],
  "total": 1,
}
`;
